# 1.Read the csv in a data_frame. (10 points)

In [51]:
import pandas as pd

In [52]:
import numpy as np

In [61]:
# Load the CSV file into a data frame
df = pd.read_csv('Book.csv')
df

Identifier             Edition Statement      Place of Publication  \
0            206                           NaN                    London   
1            216                           NaN  London; Virtue & Yorston   
2            218                           NaN                    London   
3            472                           NaN                    London   
4            480  A new edition, revised, etc.                    London   
...          ...                           ...                       ...   
8282     4158088                           NaN                    London   
8283     4158128                           NaN                     Derby   
8284     4159563                           NaN                    London   
8285     4159587                           NaN       Newcastle upon Tyne   
8286     4160339                           NaN                    London   

     Date of Publication                Publisher  \
0            1879 [1878]         S. Tinsley & Co.   
1                   1868             Virtue & Co.   
2                   1869    Bradbury, Evans & Co.   
3                   1851            James Darling   
4                   1857     Wertheim & Macintosh   
...                  ...                      ...   
8282                1838                      NaN   
8283            1831, 32          M. Mozley & Son   
8284           [1806]-22  T. Cadell and W. Davies   
8285                1834         Mackenzie & Dent   
8286             1834-43                      NaN   

                                                  Title  \
0                     Walter Forbes. [A novel.] By A. A   
1     All for Greed. [A novel. The dedication signed...   
2     Love the Avenger. By the author of “All for Gr...   
3     Welsh Sketches, chiefly ecclesiastical, to the...   
4     [The World in which I live, and my place in it...   
...                                                 ...   
8282  The Parochial History of Cornwall, founded on,...   
8283  The History and Gazetteer of the County of Der...   
8284  Magna Britannia; being a concise topographical...   
8285  An historical, topographical and descriptive v...   
8286  Collectanea Topographica et Genealogica. [Firs...   

                                                 Author  \
0                                                 A. A.   
1                                             A., A. A.   
2                                             A., A. A.   
3                                             A., E. S.   
4                                             A., E. S.   
...                                                 ...   
8282                 GIDDY, afterwards GILBERT, Davies.   
8283                         GLOVER, Stephen - of Derby   
8284  LYSONS, Daniel - M.A., F.R.S., and LYSONS (Sam...   
8285                              Mackenzie, E. (Eneas)   
8286                                                NaN   

                                           Contributors  Corporate Author  \
0                                       FORBES, Walter.               NaN   
1          BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
2          BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
3                           Appleyard, Ernest Silvanus.               NaN   
4                                   BROOME, John Henry.               NaN   
...                                                 ...               ...   
8282  BOASE, Henry Samuel.|HALS, William.|LYSONS, Da...               NaN   
8283                                     NOBLE, Thomas.               NaN   
8284           GREGSON, Matthew.|LYSONS, Samuel - F.R.S               NaN   
8285                               ROSS, M. - of Durham               NaN   
8286            BANDINEL, Bulkeley.|Nichols, John Gough               NaN   

      Corporate Contributors Former owner  Engraver Issuance type  \
0                        NaN          NaN   

# 2.Remove the following columns from the data_frame (20 points)
a.'Edition Statement', 
b.'Corporate Author', 
c.'Corporate Contributors', 
d.'Former owner', 
e.'Engraver', 
f.'Contributors', 
g.'Issuance type', 
h.'Shelfmarks'


In [54]:
# Remove unnecessary columns
df = df.drop(['Edition Statement', 'Corporate Author', 'Corporate Contributors', 'Former owner', 'Engraver', 'Contributors', 'Issuance type', 'Shelfmarks'], axis=1)

In [60]:
df

Identifier Place of Publication Date of Publication  \
0            206               London                1879   
1            216               London                1868   
2            218               London                1869   
3            472               London                1851   
4            480               London                1857   
...          ...                  ...                 ...   
8282     4158088               London                1838   
8283     4158128                Derby                1831   
8284     4159563               London                 NaN   
8285     4159587  Newcastle upon Tyne                1834   
8286     4160339               London                1834   

                    Publisher  \
0            S. Tinsley & Co.   
1                Virtue & Co.   
2       Bradbury, Evans & Co.   
3               James Darling   
4        Wertheim & Macintosh   
...                       ...   
8282                      NaN   
8283          M. Mozley & Son   
8284  T. Cadell and W. Davies   
8285         Mackenzie & Dent   
8286                      NaN   

                                                  Title  \
0                                         Walter Forbes   
1                                         All for Greed   
2                                      Love the Avenger   
3     Welsh Sketches, chiefly ecclesiastical, to the...   
4         The World in which I live, and my place in it   
...                                                 ...   
8282  The Parochial History of Cornwall, founded on,...   
8283  The History and Gazetteer of the County of Der...   
8284  Magna Britannia; being a concise topographical...   
8285  An historical, topographical and descriptive v...   
8286            Collectanea Topographica et Genealogica   

                                                 Author  \
0                                                 A. A.   
1                                             A., A. A.   
2                                             A., A. A.   
3                                             A., E. S.   
4                                             A., E. S.   
...                                                 ...   
8282                 GIDDY, afterwards GILBERT, Davies.   
8283                         GLOVER, Stephen - of Derby   
8284  LYSONS, Daniel - M.A., F.R.S., and LYSONS (Sam...   
8285                              Mackenzie, E. (Eneas)   
8286                                                NaN   

                                             Flickr URL  \
0     http://www.flickr.com/photos/britishlibrary/ta...   
1     http://www.flickr.com/photos/britishlibrary/ta...   
2     http://www.flickr.com/photos/britishlibrary/ta...   
3     http://www.flickr.com/photos/britishlibrary/ta...   
4     http://www.flickr.com/photos/britishlibrary/ta...   
...                                                 ...   
8282  http://www.flickr.com/photos/britishlibrary/ta...   
8283  http://www.flickr.com/photos/britishlibrary/ta...   
8284  http://www.flickr.com/photos/britishlibrary/ta...   
8285  http://www.flickr.com/photos/britishlibrary/ta...   
8286  http://www.flickr.com/photos/britishlibrary/ta...   

                                            first_name  last_name  
0                                                  NaN      A. A.  
1                                                A. A.         A.  
2                                                A. A.         A.  
3                                                E. S.         A.  
4                                                E. S.         A.  
...                                                ...        ...  
8282                       afterwards GILBERT, Davies.      GIDDY  
8283                                Stephen - of Derby     GLOVER  
8284  Daniel - M.A., F.R.S., and LYSONS (Samuel) F.R.S     LYSONS  
8285                                        E. (Eneas)  Mackenzie  
8

# 3.Check the content of the column- 'Date of Publication’ and define a function to clean the value. E.g., the original value: 1879 [1878] to: 1879; [1858.] to 1858; (20 points)

In [56]:
def clean_Date_of_Publication( ):
    df["Date of Publication"] = df["Date of Publication"].str.extract( r'^(\d{4})',expand = False)
    return df
clean_Date_of_Publication( )

Identifier      Place of Publication Date of Publication  \
0            206                    London                1879   
1            216  London; Virtue & Yorston                1868   
2            218                    London                1869   
3            472                    London                1851   
4            480                    London                1857   
...          ...                       ...                 ...   
8282     4158088                    London                1838   
8283     4158128                     Derby                1831   
8284     4159563                    London                 NaN   
8285     4159587       Newcastle upon Tyne                1834   
8286     4160339                    London                1834   

                    Publisher  \
0            S. Tinsley & Co.   
1                Virtue & Co.   
2       Bradbury, Evans & Co.   
3               James Darling   
4        Wertheim & Macintosh   
...                       ...   
8282                      NaN   
8283          M. Mozley & Son   
8284  T. Cadell and W. Davies   
8285         Mackenzie & Dent   
8286                      NaN   

                                                  Title  \
0                     Walter Forbes. [A novel.] By A. A   
1     All for Greed. [A novel. The dedication signed...   
2     Love the Avenger. By the author of “All for Gr...   
3     Welsh Sketches, chiefly ecclesiastical, to the...   
4     [The World in which I live, and my place in it...   
...                                                 ...   
8282  The Parochial History of Cornwall, founded on,...   
8283  The History and Gazetteer of the County of Der...   
8284  Magna Britannia; being a concise topographical...   
8285  An historical, topographical and descriptive v...   
8286  Collectanea Topographica et Genealogica. [Firs...   

                                                 Author  \
0                                                 A. A.   
1                                             A., A. A.   
2                                             A., A. A.   
3                                             A., E. S.   
4                                             A., E. S.   
...                                                 ...   
8282                 GIDDY, afterwards GILBERT, Davies.   
8283                         GLOVER, Stephen - of Derby   
8284  LYSONS, Daniel - M.A., F.R.S., and LYSONS (Sam...   
8285                              Mackenzie, E. (Eneas)   
8286                                                NaN   

                                             Flickr URL  
0     http://www.flickr.com/photos/britishlibrary/ta...  
1     http://www.flickr.com/photos/britishlibrary/ta...  
2     http://www.flickr.com/photos/britishlibrary/ta...  
3     http://www.flickr.com/photos/britishlibrary/ta...  
4     http://www.flickr.com/photos/britishlibrary/ta...  
...                                                 ...  
8282  http://www.flickr.com/photos/britishlibrary/ta...  
8283  http://www.flickr.com/photos/britishlibrary/ta...  
8284  http://www.flickr.com/photos/britishlibrary/ta...  
8285  http://www.flickr.com/photos/britishlibrary/ta...  
8286  http://www.flickr.com/photos/britishlibrary/ta...  

[8287 rows x 7 columns]

# 4.Check the content of the column- ' Author’ and define a function to clean the value. And split the name to first name and last name. (10 points)

In [57]:
def clean_author ():
    df['first_name']=df['Author'].str.split(',',1).str[1].str.strip()
    df['last_name']=df['Author'].str.split(',',1).str[0].str.strip()
    return df
clean_author()

Identifier      Place of Publication Date of Publication  \
0            206                    London                1879   
1            216  London; Virtue & Yorston                1868   
2            218                    London                1869   
3            472                    London                1851   
4            480                    London                1857   
...          ...                       ...                 ...   
8282     4158088                    London                1838   
8283     4158128                     Derby                1831   
8284     4159563                    London                 NaN   
8285     4159587       Newcastle upon Tyne                1834   
8286     4160339                    London                1834   

                    Publisher  \
0            S. Tinsley & Co.   
1                Virtue & Co.   
2       Bradbury, Evans & Co.   
3               James Darling   
4        Wertheim & Macintosh   
...                       ...   
8282                      NaN   
8283          M. Mozley & Son   
8284  T. Cadell and W. Davies   
8285         Mackenzie & Dent   
8286                      NaN   

                                                  Title  \
0                     Walter Forbes. [A novel.] By A. A   
1     All for Greed. [A novel. The dedication signed...   
2     Love the Avenger. By the author of “All for Gr...   
3     Welsh Sketches, chiefly ecclesiastical, to the...   
4     [The World in which I live, and my place in it...   
...                                                 ...   
8282  The Parochial History of Cornwall, founded on,...   
8283  The History and Gazetteer of the County of Der...   
8284  Magna Britannia; being a concise topographical...   
8285  An historical, topographical and descriptive v...   
8286  Collectanea Topographica et Genealogica. [Firs...   

                                                 Author  \
0                                                 A. A.   
1                                             A., A. A.   
2                                             A., A. A.   
3                                             A., E. S.   
4                                             A., E. S.   
...                                                 ...   
8282                 GIDDY, afterwards GILBERT, Davies.   
8283                         GLOVER, Stephen - of Derby   
8284  LYSONS, Daniel - M.A., F.R.S., and LYSONS (Sam...   
8285                              Mackenzie, E. (Eneas)   
8286                                                NaN   

                                             Flickr URL  \
0     http://www.flickr.com/photos/britishlibrary/ta...   
1     http://www.flickr.com/photos/britishlibrary/ta...   
2     http://www.flickr.com/photos/britishlibrary/ta...   
3     http://www.flickr.com/photos/britishlibrary/ta...   
4     http://www.flickr.com/photos/britishlibrary/ta...   
...                                                 ...   
8282  http://www.flickr.com/photos/britishlibrary/ta...   
8283  http://www.flickr.com/photos/britishlibrary/ta...   
8284  http://www.flickr.com/photos/britishlibrary/ta...   
8285  http://www.flickr.com/photos/britishlibrary/ta...   
8286  http://www.flickr.com/photos/britishlibrary/ta...   

                                            first_name  last_name  
0                                                  NaN      A. A.  
1                                                A. A.         A.  
2                                                A. A.         A.  
3                                                E. S.         A.  
4                                                E. S.         A.  
...                                                ...        ...  
8282                       afterwards GILBERT, Davies.      GIDDY  
8283                                Stephen - of Derby     GLOVER  
8284  Daniel - M.A., F.R.S., and LYSONS (Samuel) F.R.S     LYSONS  
8285     

# 5.	Check the content of the column- ‘Title’ and define a function to clean the value. E.g., the original value: Walter Forbes. [A novel.] By A. A to: Walter Forbes;  Love the Avenger. By the author of “All for Gr.. to Love The Avenger. (10 points)

In [58]:
def clean_title():
    df['Title'] = df['Title'].str.strip('[')
    df['Title'] = df['Title'].str.split('.',1).str[0].str.strip()
    return df
clean_title()

Identifier      Place of Publication Date of Publication  \
0            206                    London                1879   
1            216  London; Virtue & Yorston                1868   
2            218                    London                1869   
3            472                    London                1851   
4            480                    London                1857   
...          ...                       ...                 ...   
8282     4158088                    London                1838   
8283     4158128                     Derby                1831   
8284     4159563                    London                 NaN   
8285     4159587       Newcastle upon Tyne                1834   
8286     4160339                    London                1834   

                    Publisher  \
0            S. Tinsley & Co.   
1                Virtue & Co.   
2       Bradbury, Evans & Co.   
3               James Darling   
4        Wertheim & Macintosh   
...                       ...   
8282                      NaN   
8283          M. Mozley & Son   
8284  T. Cadell and W. Davies   
8285         Mackenzie & Dent   
8286                      NaN   

                                                  Title  \
0                                         Walter Forbes   
1                                         All for Greed   
2                                      Love the Avenger   
3     Welsh Sketches, chiefly ecclesiastical, to the...   
4         The World in which I live, and my place in it   
...                                                 ...   
8282  The Parochial History of Cornwall, founded on,...   
8283  The History and Gazetteer of the County of Der...   
8284  Magna Britannia; being a concise topographical...   
8285  An historical, topographical and descriptive v...   
8286            Collectanea Topographica et Genealogica   

                                                 Author  \
0                                                 A. A.   
1                                             A., A. A.   
2                                             A., A. A.   
3                                             A., E. S.   
4                                             A., E. S.   
...                                                 ...   
8282                 GIDDY, afterwards GILBERT, Davies.   
8283                         GLOVER, Stephen - of Derby   
8284  LYSONS, Daniel - M.A., F.R.S., and LYSONS (Sam...   
8285                              Mackenzie, E. (Eneas)   
8286                                                NaN   

                                             Flickr URL  \
0     http://www.flickr.com/photos/britishlibrary/ta...   
1     http://www.flickr.com/photos/britishlibrary/ta...   
2     http://www.flickr.com/photos/britishlibrary/ta...   
3     http://www.flickr.com/photos/britishlibrary/ta...   
4     http://www.flickr.com/photos/britishlibrary/ta...   
...                                                 ...   
8282  http://www.flickr.com/photos/britishlibrary/ta...   
8283  http://www.flickr.com/photos/britishlibrary/ta...   
8284  http://www.flickr.com/photos/britishlibrary/ta...   
8285  http://www.flickr.com/photos/britishlibrary/ta...   
8286  http://www.flickr.com/photos/britishlibrary/ta...   

                                            first_name  last_name  
0                                                  NaN      A. A.  
1                                                A. A.         A.  
2                                                A. A.         A.  
3                                                E. S.         A.  
4                                                E. S.         A.  
...                                                ...        ...  
8282                       afterwards GILBERT, Davies.      GIDDY  
8283                                Stephen - of Derby     GLOVER  
8284  Daniel - M.A., F.R.S., and LYSONS (Samuel) F.R.S     LYSONS  
8285     

# 6.	Check the content of the column- ‘Place of Publication’ and define a function to clean the value. E.g., the original value: London; Virtue & Yorston to: London. (10 points)

In [59]:
def clean_place_of_publication():
    df['Place of Publication'] = np.where(df['Place of Publication'].str.contains('London'), 'London',
                                           df['Place of Publication'].str.replace('-', ' '))
    return df
clean_place_of_publication()



Identifier Place of Publication Date of Publication  \
0            206               London                1879   
1            216               London                1868   
2            218               London                1869   
3            472               London                1851   
4            480               London                1857   
...          ...                  ...                 ...   
8282     4158088               London                1838   
8283     4158128                Derby                1831   
8284     4159563               London                 NaN   
8285     4159587  Newcastle upon Tyne                1834   
8286     4160339               London                1834   

                    Publisher  \
0            S. Tinsley & Co.   
1                Virtue & Co.   
2       Bradbury, Evans & Co.   
3               James Darling   
4        Wertheim & Macintosh   
...                       ...   
8282                      NaN   
8283          M. Mozley & Son   
8284  T. Cadell and W. Davies   
8285         Mackenzie & Dent   
8286                      NaN   

                                                  Title  \
0                                         Walter Forbes   
1                                         All for Greed   
2                                      Love the Avenger   
3     Welsh Sketches, chiefly ecclesiastical, to the...   
4         The World in which I live, and my place in it   
...                                                 ...   
8282  The Parochial History of Cornwall, founded on,...   
8283  The History and Gazetteer of the County of Der...   
8284  Magna Britannia; being a concise topographical...   
8285  An historical, topographical and descriptive v...   
8286            Collectanea Topographica et Genealogica   

                                                 Author  \
0                                                 A. A.   
1                                             A., A. A.   
2                                             A., A. A.   
3                                             A., E. S.   
4                                             A., E. S.   
...                                                 ...   
8282                 GIDDY, afterwards GILBERT, Davies.   
8283                         GLOVER, Stephen - of Derby   
8284  LYSONS, Daniel - M.A., F.R.S., and LYSONS (Sam...   
8285                              Mackenzie, E. (Eneas)   
8286                                                NaN   

                                             Flickr URL  \
0     http://www.flickr.com/photos/britishlibrary/ta...   
1     http://www.flickr.com/photos/britishlibrary/ta...   
2     http://www.flickr.com/photos/britishlibrary/ta...   
3     http://www.flickr.com/photos/britishlibrary/ta...   
4     http://www.flickr.com/photos/britishlibrary/ta...   
...                                                 ...   
8282  http://www.flickr.com/photos/britishlibrary/ta...   
8283  http://www.flickr.com/photos/britishlibrary/ta...   
8284  http://www.flickr.com/photos/britishlibrary/ta...   
8285  http://www.flickr.com/photos/britishlibrary/ta...   
8286  http://www.flickr.com/photos/britishlibrary/ta...   

                                            first_name  last_name  
0                                                  NaN      A. A.  
1                                                A. A.         A.  
2                                                A. A.         A.  
3                                                E. S.         A.  
4                                                E. S.         A.  
...                                                ...        ...  
8282                       afterwards GILBERT, Davies.      GIDDY  
8283                                Stephen - of Derby     GLOVER  
8284  Daniel - M.A., F.R.S., and LYSONS (Samuel) F.R.S     LYSONS  
8285                                        E. (Eneas)  Mackenzie  
8